File: Semester5/Formal Languages and Compiler Design/Tema2/LexicalAnalyzer/HashTable.py
class HashTable:
    def __init__(self, size=100):
        self.size = size
        self.table = [None] * size
    
    def _hash(self, key):
        return (sum(ord(c) for c in key) + len(key)) % self.size

    def put(self, key):
        index = self._hash(key)

        original_index = index
        while self.table[index] is not None:
            if self.table[index] == key:
                return index
            index = (index + 1) % self.size
            if index == original_index:
                return

        self.table[index] = key
        return index

    def get(self, key):
        index = self._hash(key)
        original_index = index
        while self.table[index] is not None:
            if self.table[index] == key:
                return index
            index = (index + 1) % self.size
            if index == original_index:
                break
        return None
    
    def display(self):
        print("Symbol Table (TS):")
        for i, entry in enumerate(self.table):
            if entry is not None:
                print(f"Index {i}: {entry}")

File: Semester5/Formal Languages and Compiler Design/Tema3/HashTable.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include "HashTable.h"

unsigned int hash(const char *key) {
    unsigned int hash_value = 0;
    for (int i = 0; key[i] != '\0'; i++) {
        hash_value += key[i];
    }
    hash_value += strlen(key);
    return hash_value % TABLE_SIZE;
}

HashTable* create_table() {
    HashTable *hash_table = malloc(sizeof(HashTable));
    for (int i = 0; i < TABLE_SIZE; i++) {
        hash_table->table[i] = NULL;
    }
    return hash_table;
}

int put(HashTable *hash_table, const char *key) {
    unsigned int index = hash(key);
    unsigned int original_index = index;

    while (hash_table->table[index] != NULL) {
        if (strcmp(hash_table->table[index]->key, key) == 0) {
            return index;
        }
        index = (index + 1) % TABLE_SIZE;
        if (index == original_index) {
            return -1;
        }
    }

    hash_table->table[index] = malloc(sizeof(HashEntry));
    hash_table->table[index]->key = strdup(key);
    return index;
}

int get(HashTable *hash_table, const char *key) {
    unsigned int index = hash(key);
    unsigned int original_index = index;

    while (hash_table->table[index] != NULL) {
        if (strcmp(hash_table->table[index]->key, key) == 0) {
            return index;
        }
        index = (index + 1) % TABLE_SIZE;
        if (index == original_index) {
            break;
        }
    }

    return -1;
}

void free_table(HashTable *hash_table) {
    for (int i = 0; i < TABLE_SIZE; i++) {
        if (hash_table->table[i] != NULL) {
            free(hash_table->table[i]->key);
            free(hash_table->table[i]);
        }
    }
    free(hash_table);
}